### DISCUSSION TITLE
Parallel fetch of sources (rastersources)?

### DISCUSSION THREAD

**Discussion by pcace** [maplibre/maplibre-gl-js#4759] [Q&A]
Hi there, 

first of all i am not sure if its a bug or by design. I have noticed some quite bad performance when using many raster sources which have the same address. i currently add sources like this: 

```typescript
<Source
          key={`${layer.id}_${layer.name}_${layer.source.type}`}
          id={`${layer.id}_${layer.name}_${layer.source.type}`}
          type="raster"
          tiles={[
            `${process.env.SELFSERVING_TILESERVER}/wmstoxyz/{z}/{x}/{y}?useCache=true&tileSize=${TILESIZE}&address=${url}`,
          ]}
          attribution={layer.source.attribution}
          tileSize={TILESIZE}
        >
          {layer.mapStyles?.map((mapStyle) => {
            return <Layer beforeId={beforeId} key={mapStyle.id} {...mapStyle} />
          })}
        </Source>
```

for the tiles url i use some kind of a tile - proxy: its a fastify server which takes an url, and converts tiles to the proper projection. the fastify server can handle multiple parallel queries (tested). 

BUT: maplibre seems to work on each tile sequentially. this seems to be very unperformant, since it waits for each tile to be finished after working at the next tile. in chrome network tab it looks like this: 

![image](https://github.com/user-attachments/assets/3f00fbf1-e287-4c99-bc8e-10c01c734967)

this clearly shows that every tile gets queried sequentially. 
i have the feeling this especially happens if the tiles are coming from the same address (whcih in my case all tiles are because they come from the "proxy").

can that be? 
Is that intentionally? Any idea to make that fast(er)?

Thanks a lot for ideas, 

pcace

**Comment by HarelM** [comment]
Thanks for taking the time to report this issue.
I think you should check your configuration.
Images are fetched in parallel as can be seen in the following file:
https://github.com/maplibre/maplibre-gl-js/blob/main/src/util/image_request.ts
Images should be fetched in parallel up to 16 as can be seen in the following configuration:
https://github.com/maplibre/maplibre-gl-js/blob/main/src/util/config.ts#L24
Also browser limit is usually 6 parallel requests (to any address), so it might be that other requests are being made at the same time.
